In [26]:
!pip install -q pyomo

In [27]:
from pyomo.environ import *

In [28]:
import numpy as np

In [29]:
import pandas as pd


In this lab, we will consider non-linear problems which can be converted to linear programs. One famous example is the absolute value function given as: 
\begin{align}
|x| = \begin{cases}
        x  & \text{ if } x\geq 0 \\ 
        -x & \text{ otherwise }.
      \end{cases}
\end{align}


We will now consider the absolute value function appearing in the objective function. Consider the optimization problem $\textbf{(OP)}$:
\begin{align}
\min \ |x_1| + |x_2| + |x_3| + |x_4| + |x_5| &  \nonumber \\
{\rm{s.t.}} \ 85 x_1 + 92 x_2 + 45 x_3 + 27 x_4 + 31 x_5 & \geq 1 \nonumber \\
92 x_1 + 54 x_2 + 22 x_3 + 20 x_4 + 7 x_5 & \geq 1 \nonumber \\
96 x_1 + 67 x_2 + 29 x_3 + 20 x_4 + 11 x_5 & \geq 1 \nonumber \\
-91 x_1 - 57 x_2 - 33 x_3 - 23 x_4 - 12 x_5 & \geq 1 \nonumber \\
-99 x_1 - 75 x_2 - 26 x_3 - 24 x_4 - 41 x_5 & \geq 1 \nonumber \\
-98 x_1 - 99 x_2 - 57 x_3 - 45 x_4 - 65 x_5 & \geq 1 \nonumber
\end{align}

In this optimization problem $\textbf{(OP)}$, we wish to minimize the sum of absolute values of the decision variables. Such problems are useful when the decision variables might take both positive and negative values, and we want to optimize the magnitudes of the decision variables. 

We will now consider two possible ways to solve this problem. 

$\textbf{Approach 1:}$ 

We can use a substitution of the absolute function values of the variables to new variables as $u_i = |x_i|, \ \forall i \in \{1,\ldots,5\}$. With this substitution, note that we have $u_i\geq 0, \ \forall i \in \{1,\ldots,5\}$. Also, the following inequalities are satisfied: $x_i \leq u_i, -x_i \leq u_i, \ \forall i \in \{1,\ldots,5\}$. Thus we can transform the problem $\textbf{(OP)}$ as the following optimization problem $\textbf{(OP1)}$, where the absolute values of $x_i$ are replaced with $u_i$ and constraints related to the new variables $u_i$ are introduced. Thus we have the problem $\textbf{(OP1)}$:

\begin{align}
\min \ u_1 + u_2 + u_3 + u_4 + u_5 &  \nonumber \\
{\rm{s.t.}} \ 85 x_1 + 92 x_2 + 45 x_3 + 27 x_4 + 31 x_5 & \geq 1 \nonumber \\
92 x_1 + 54 x_2 + 22 x_3 + 20 x_4 + 7 x_5 & \geq 1 \nonumber \\
96 x_1 + 67 x_2 + 29 x_3 + 20 x_4 + 11 x_5 & \geq 1 \nonumber \\
-91 x_1 - 57 x_2 - 33 x_3 - 23 x_4 - 12 x_5 & \geq 1 \nonumber \\
-99 x_1 - 75 x_2 - 26 x_3 - 24 x_4 - 41 x_5 & \geq 1 \nonumber \\
-98 x_1 - 99 x_2 - 57 x_3 - 45 x_4 - 65 x_5 & \geq 1 \nonumber \\
u_i \geq x_i, & \  \forall i \in \{1,\ldots,5\} \nonumber \\ 
u_i \geq -x_i, & \  \forall i \in \{1,\ldots,5\} \nonumber \\ 
u_i \geq 0, & \ \forall i \in \{1,\ldots,5\}. 
\end{align}

Note that the number of decision variables in $\textbf{(OP1)}$ is twice the number of decision variables in $\textbf{(OP)}$. However the objective function and the constraints in $\textbf{(OP1)}$ are now linear, and thus $\textbf{(OP1)}$ is a linear program. 




In [30]:
coef = np.loadtxt('lab6_practice_coef.txt', delimiter=',')

In [31]:
print(coef.shape)
print('Num rows:',coef.shape[0])
print('Num cols:',coef.shape[1])

(17, 11)
Num rows: 17
Num cols: 11


In [32]:
print(coef)

[[  0.   0.   0.   0.   0.   1.   1.   1.   1.   1.   0.]
 [ 85.  92.  45.  27.  31.   0.   0.   0.   0.   0.   1.]
 [ 92.  54.  22.  20.   7.   0.   0.   0.   0.   0.   1.]
 [ 96.  67.  29.  20.  11.   0.   0.   0.   0.   0.   1.]
 [-91. -57. -33. -23. -12.   0.   0.   0.   0.   0.   1.]
 [-99. -75. -26. -24. -41.   0.   0.   0.   0.   0.   1.]
 [-98. -99. -57. -45. -65.   0.   0.   0.   0.   0.   1.]
 [ -1.   0.   0.   0.   0.   1.   0.   0.   0.   0.   0.]
 [  0.  -1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.   0.  -1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.   0.  -1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.   0.  -1.   0.   0.   0.   0.   1.   0.]
 [  1.   0.   0.   0.   0.   1.   0.   0.   0.   0.   0.]
 [  0.   1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.   0.   1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.   0.   1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.   0.   1.   0.   0.   0.   0.   1.   0.]]


In [33]:
model=ConcreteModel()

In [34]:
N=coef.shape[1]-1

In [35]:
M=coef.shape[0]-1

In [36]:
obj_coef = coef[0,:-1]
print(obj_coef.shape)
print(obj_coef)

(10,)
[0. 0. 0. 0. 0. 1. 1. 1. 1. 1.]


In [37]:
constr_coef = coef[1:,:-1]
print(constr_coef.shape)
print(constr_coef)

(16, 10)
[[ 85.  92.  45.  27.  31.   0.   0.   0.   0.   0.]
 [ 92.  54.  22.  20.   7.   0.   0.   0.   0.   0.]
 [ 96.  67.  29.  20.  11.   0.   0.   0.   0.   0.]
 [-91. -57. -33. -23. -12.   0.   0.   0.   0.   0.]
 [-99. -75. -26. -24. -41.   0.   0.   0.   0.   0.]
 [-98. -99. -57. -45. -65.   0.   0.   0.   0.   0.]
 [ -1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.  -1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.  -1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.  -1.   0.   0.   0.   0.   1.   0.]
 [  0.   0.   0.   0.  -1.   0.   0.   0.   0.   1.]
 [  1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.   1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.   1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.   1.   0.   0.   0.   0.   1.   0.]
 [  0.   0.   0.   0.   1.   0.   0.   0.   0.   1.]]


In [38]:
constr_rhs = coef[1:,-1]
print(constr_rhs.shape)
print(constr_rhs)

(16,)
[1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [39]:
col_indices_x = np.arange(int(N/2))
print(col_indices_x)

[0 1 2 3 4]


In [40]:
model.x = Var(col_indices_x)

In [41]:
col_indices_u = np.arange(int(N/2))
print(col_indices_u)

[0 1 2 3 4]


In [42]:
model.u = Var(col_indices_u, domain=NonNegativeReals)

In [43]:
row_indices = np.arange(M)
print(row_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]


In [44]:
model.constraints = ConstraintList()

In [45]:
for i in row_indices:
  model.constraints.add(summation(constr_coef[i][0:int(N/2)],model.x) + summation(constr_coef[i][int(N/2):N],model.u) >= constr_rhs[i])

In [46]:
model.objective = Objective(expr = summation(obj_coef[0:int(N/2)],model.x) + summation(obj_coef[int(N/2):N],model.u), sense=minimize)

In [47]:
model.pprint()

3 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    u_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

2 Var Declarations
    u : Size=5, Index=u_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeReals
          1 :     0 :  None :  None : False :  True : NonNegativeReals
          2 :     0 :  None :  None : False :  True : NonNegativeReals
          3 :     0 :  None :  None : False :  True : NonNegativeReals
          4 :     0 :  None :  None : False :  True : NonNegativeReals
    x : Siz

In [48]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 148492 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [49]:
opt_cbc=SolverFactory('cbc')

In [50]:
result1=opt_cbc.solve(model)

#Que.2

In [51]:
print('Solver status:', result1.solver.status)
print('Solver termination condition:',result1.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


The Solver status is ok and the termination condition is Optimal

In [52]:
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for j in col_indices_x:
    print('x[',j,']:', model.x[j].value)

for j in col_indices_u:
    print('u[',j,']:', model.u[j].value)

print('\nConstraints')
model.constraints.display()


Objective =  0.547955725

Decision Variables
x[ 0 ]: -0.046448575
x[ 1 ]: 0.18612441
x[ 2 ]: -0.17131802
x[ 3 ]: 0.0
x[ 4 ]: -0.14406472
u[ 0 ]: 0.046448575
u[ 1 ]: 0.18612441
u[ 2 ]: 0.17131802
u[ 3 ]: 0.0
u[ 4 ]: 0.14406472

Constraints
constraints : Size=16
    Key : Lower : Body               : Upper
      1 :   1.0 : 0.9999996250000001 :  None
      2 :   1.0 : 0.9999997599999997 :  None
      3 :   1.0 :  1.458337769999999 :  None
      4 :   1.0 : 1.0000002550000002 :  None
      5 :   1.0 :  1.000000215000001 :  None
      6 :   1.0 : 5.2549776999999995 :  None
      7 :   0.0 :         0.09289715 :  None
      8 :   0.0 :                0.0 :  None
      9 :   0.0 :         0.34263604 :  None
     10 :   0.0 :                0.0 :  None
     11 :   0.0 :         0.28812944 :  None
     12 :   0.0 :                0.0 :  None
     13 :   0.0 :         0.37224882 :  None
     14 :   0.0 :                0.0 :  None
     15 :   0.0 :                0.0 :  None
     16 :   0.0 : 

Yes,The solver yields an optimal solution.

#Que.3

The objective Function value is:0.547955725

The values of decision variales are:
$$
x_1= -0.046448575\\
x_2= 0.18612441\\
x_3= -0.17131802\\
x_4= 0.0\\
x_5= -0.14406472\\
u_1= 0.046448575\\
u_2= 0.18612441\\
u_3= 0.17131802\\
u_4= 0.0\\
u_5= 0.14406472 $$ 

Contraints 1,2,4,5,8,10,12,14,15 and 16 are active as they hold at equality.
and rest constraints are not active as they don't hold at equality.



Recall our optimization problem $\textbf{(OP)}$:
\begin{align}
\min \ |x_1| + |x_2| + |x_3| + |x_4| + |x_5| &  \nonumber \\
{\rm{s.t.}} \ 85 x_1 + 92 x_2 + 45 x_3 + 27 x_4 + 31 x_5 & \geq 1 \nonumber \\
92 x_1 + 54 x_2 + 22 x_3 + 20 x_4 + 7 x_5 & \geq 1 \nonumber \\
96 x_1 + 67 x_2 + 29 x_3 + 20 x_4 + 11 x_5 & \geq 1 \nonumber \\
-91 x_1 - 57 x_2 - 33 x_3 - 23 x_4 - 12 x_5 & \geq 1 \nonumber \\
-99 x_1 - 75 x_2 - 26 x_3 - 24 x_4 - 41 x_5 & \geq 1 \nonumber \\
-98 x_1 - 99 x_2 - 57 x_3 - 45 x_4 - 65 x_5 & \geq 1 \nonumber
\end{align}

Now we will consider a different approach to make the problem $\textbf{(OP)}$ linear. 

$\textbf{Approach 2:}$

In this approach we will use some fundamental properties of real numbers. Indeed, for any real number $x$, we can write $x=a-b$ where $a\geq 0, b\geq 0$. 

Hence we can substitute $|x| = |a-b|, \ a\geq 0, b\geq 0$. 

Now we will consider another interesting property of real numbers: 
$|u-v| = u+v \iff u\geq 0, v\geq 0, uv = 0$. 

Using this property we can write $|x| = a+b, \ a\geq 0, \ b\geq 0$. Note that this replacement will imply that $ab=0$. 

Thus we can transform the optimization problem $\textbf{(OP)}$ into a new optimization problem $\textbf{(OP2)}$ where $x_i = a_i - b_i, \text{ and } |x_i| = a_i + b_i, \ \forall i \in \{1,\ldots,5\}$ where $a_i \geq 0, b_i \geq 0, \forall i \in \{1,\ldots,5\}$. Thus we have the new optimization problem $\textbf{(OP2)}$: 

\begin{align}
\min \ (a_1 + b_1) +  (a_2 + b_2) + (a_3 + b_3) + (a_4 + b_4) + (a_5 + b_5) &  \nonumber \\
{\rm{s.t.}} \ 85 (a_1-b_1) + 92 (a_2-b_2) + 45 (a_3-b_3) + 27 (a_4-b_4) + 31 (a_5-b_5) & \geq 1 \nonumber \\
92 (a_1-b_1) + 54 (a_2-b_2) + 22 (a_3-b_3) + 20 (a_4-b_4) + 7 (a_5-b_5) & \geq 1 \nonumber \\
96 (a_1-b_1) + 67 (a_2-b_2) + 29 (a_3-b_3) + 20 (a_4-b_4) + 11 (a_5-b_5) & \geq 1 \nonumber \\
-91 (a_1-b_1) - 57 (a_2-b_2) - 33 (a_3-b_3) - 23 (a_4-b_4) - 12 (a_5-b_5) & \geq 1 \nonumber \\
-99 (a_1-b_1) - 75 (a_2-b_2) - 26 (a_3-b_3) - 24 (a_4-b_4) - 41 (a_5-b_5) & \geq 1 \nonumber \\
-98 (a_1-b_1) - 99 (a_2-b_2) - 57 (a_3-b_3) - 45 (a_4-b_4) - 65 (a_5-b_5) & \geq 1 \nonumber \\ 
a_i \geq 0, \ b_i \geq 0, & \ \forall i \in \{1,\ldots,5\}. \nonumber 
\end{align}


In [54]:
coef_op2=np.loadtxt('lab6_practice_coef_0P2.txt',delimiter=',')

In [55]:
print(coef_op2.shape)
print('Number of rows:',coef_op2.shape[0])
print('Number of columns:',coef_op2.shape[1])

(7, 11)
Number of rows: 7
Number of columns: 11


In [56]:
print(coef_op2)

[[  1.   1.   1.   1.   1.   1.   1.   1.   1.   1.   0.]
 [ 85.  92.  45.  27.  31. -85. -92. -45. -27. -31.   1.]
 [ 92.  54.  22.  20.   7. -92. -54. -22. -20.  -7.   1.]
 [ 96.  67.  29.  20.  11. -96. -67. -29. -20. -11.   1.]
 [-91. -57. -33. -23. -12.  91.  57.  33.  23.  12.   1.]
 [-99. -75. -26. -24. -41.  99.  75.  26.  24.  41.   1.]
 [-98. -99. -57. -45. -65.  98.  99.  57.  45.  65.   1.]]


In [59]:
N=coef_op2.shape[1]-1
M=coef_op2.shape[0]-1
obj_coef_op2=coef_op2[0,:-1]

In [60]:
print(obj_coef_op2)

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


In [61]:
constr_coef_op2=coef_op2[1:,:-1]

In [62]:
print(constr_coef_op2)

[[ 85.  92.  45.  27.  31. -85. -92. -45. -27. -31.]
 [ 92.  54.  22.  20.   7. -92. -54. -22. -20.  -7.]
 [ 96.  67.  29.  20.  11. -96. -67. -29. -20. -11.]
 [-91. -57. -33. -23. -12.  91.  57.  33.  23.  12.]
 [-99. -75. -26. -24. -41.  99.  75.  26.  24.  41.]
 [-98. -99. -57. -45. -65.  98.  99.  57.  45.  65.]]


In [63]:
constr_rhs_op2=coef_op2[1:,-1]

In [64]:
model_op2=ConcreteModel()

In [65]:
col_indices_a=np.arange(int(N/2))
print(col_indices_a)

[0 1 2 3 4]


In [67]:
col_indices_b=np.arange(int(N/2))
print(col_indices_b)

[0 1 2 3 4]


In [68]:
model_op2.a=Var(col_indices_a,domain=NonNegativeReals)
model_op2.b=Var(col_indices_b,domain=NonNegativeReals)

In [69]:
row_indices_op2=np.arange(M)
print(row_indices_op2)

[0 1 2 3 4 5]


In [70]:
model_op2.constraints=ConstraintList()

In [72]:
model_op2.objective=Objective(expr=summation(obj_coef_op2[0:int(N/2)],model_op2.a)+summation(obj_coef_op2[int(N/2):N],model_op2.b), sense=minimize)

In [73]:
for i in row_indices_op2:
  model_op2.constraints.add(summation(constr_coef_op2[i][0:int(N/2)],model_op2.a) + summation(constr_coef_op2[i][int(N/2):N],model_op2.b)  >= constr_rhs_op2[i])

In [74]:
model_op2.pprint()

3 Set Declarations
    a_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    b_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}

2 Var Declarations
    a : Size=5, Index=a_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeReals
          1 :     0 :  None :  None : False :  True : NonNegativeReals
          2 :     0 :  None :  None : False :  True : NonNegativeReals
          3 :     0 :  None :  None : False :  True : NonNegativeReals
          4 :     0 :  None :  None : False :  True : NonNegativeReals
    b : Size=5, Index=b_index
        Key : Lowe

In [75]:
result2=opt_cbc.solve(model_op2)
print('Solver status:', result2.solver.status)
print('Solver termination condition:',result2.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


#Que.9

The Solver status for OP2 is ok and termination condition is Optimal.and yes the solver yields an optimal solution.

#Que.10

In [76]:
print('\nObjective = ', model_op2.objective())

print('\nDecision Variables')
for j in col_indices_a:
    print('a[',j,']:', model_op2.a[j].value)

for j in col_indices_b:
    print('b[',j,']:', model_op2.b[j].value)

print('\nConstraints')
model_op2.constraints.display()


Objective =  0.547955725

Decision Variables
a[ 0 ]: 0.0
a[ 1 ]: 0.18612441
a[ 2 ]: 0.0
a[ 3 ]: 0.0
a[ 4 ]: 0.0
b[ 0 ]: 0.046448575
b[ 1 ]: 0.0
b[ 2 ]: 0.17131802
b[ 3 ]: 0.0
b[ 4 ]: 0.14406472

Constraints
constraints : Size=6
    Key : Lower : Body               : Upper
      1 :   1.0 :  0.999999625000001 :  None
      2 :   1.0 : 0.9999997599999997 :  None
      3 :   1.0 :         1.45833777 :  None
      4 :   1.0 : 1.0000002549999998 :  None
      5 :   1.0 :        1.000000215 :  None
      6 :   1.0 :  5.254977699999998 :  None


The value of the Objective function is:0.547955725

The values of the decision variables are:
$a_1= 0.0\\
a_2= 0.18612441\\
a_3= 0.0\\
a_4= 0.0\\
a_5= 0.0\\
b_1= 0.046448575\\
b_2= 0.0\\
b_3= 0.17131802\\
b_4= 0.0\\
b_5= 0.14406472\\ $

Constraints 1,2,4 and 5 are active as they hold as equality,rest constraints are not active as they don't hold at equality.

#Que.11

In [77]:
for i in col_indices_a:
  print(f"The value of x[{i+1}] is:{model_op2.a[i].value-model_op2.b[i].value}")

The value of x[1] is:-0.046448575
The value of x[2] is:0.18612441
The value of x[3] is:-0.17131802
The value of x[4] is:0.0
The value of x[5] is:-0.14406472


#Que.12

The value of $x_i$'s obtained from **OP2** are exactly same with the values obtained from **OP1**.